[14.0][FIX] shopfloor_reception: Fix qty todo computation

mmequignon opened the pull request:
We had a concurrent update introduced here https://github.com/OCA/wms/pull/672

Updating product uom qties were causing issues.
This is basically fixed by modifying the product_qty in responses instead of updating the move lines qties.

mmequignon committed:
shopfloor_reception: Fix qty_done computation

jbaudoux commented on shopfloor_reception/services/reception.py:
```diff
@@ -766,25 +766,23 @@ def _align_product_uom_qties(self, move):
-
+        move = line.move_id
         qty_todo = move.product_uom_qty
         qty_done = sum(move.move_line_ids.mapped("qty_done"))
```
```suggestion
        qty_done = sum(
                    self.env['uom.uom'].browse(product_uom_id)._compute_quantity(qty_done, line.product_uom_id, round=False)
                     for product_uom_id, qty_done in move.move_line_ids)
                   )
```

jbaudoux commented on shopfloor_reception/services/reception.py:
```diff
@@ -766,25 +766,23 @@ def _align_product_uom_qties(self, move):
         #     return
-
+        move = line.move_id
         qty_todo = move.product_uom_qty
```
Compute all in line uom
```suggestion
        qty_todo = self.env['uom.uom'].browse(move.product_uom_id)._compute_quantity(move.product_uom_qty, move.product_uom_id, round=False)
```

jbaudoux reviewed (commented):
What about the uom?

mmequignon committed:
shopfloor_reception: Split moves before posting

mmequignon committed:
shopfloor_reception: Fix uom computation

mmequignon committed:
shopfloor_reception: Make select picking by product optional

mmequignon committed:
shopfloor_reception: set_quantity - Add cancel button

mt-software-de commented on shopfloor_reception/services/reception.py:
```diff
@@ -1235,6 +1243,11 @@ def _post_shopfloor_created_line(self, selected_line):
+        # then selected_line is the only one related to this move.
+        # In such case, we must ensure there's another move line with the remaining
+        # quantity to do, so selected_line is extracted in a new move as expected.
+        selected_line._split_partial_quantity()
```
```suggestion
        new_move_line = selected_line._split_partial_quantity()
```
We should call unmark_move_line_as_picked and set the location_dest_id back to stock.move's location_dest_id
if a split was happening 
```
            stock = self._actions_for("stock")
            stock.unmark_move_line_as_picked(new_move_line)
            new_move_line.location_dest_id = new_move_line.move_id.location_dest_id
```

mt-software-de replied:
https://github.com/camptocamp/wms/pull/118

mmequignon replied:
Hello @mt-software-de , I updated the branch, can you please rebase? I'm checking this.
Thanks!

jbaudoux replied:
This is correct but you also need to recompute the putaway.
It is easier if you always set product_uom_qty to 0, then there is no move line backorder (i.e. nothing to split).
I would do:
```
if product_uom_qty:
    selected_line.product_uom_qty = 0
```
and drop the `_split_partial_quantity`

mt-software-de replied:
> Hello @mt-software-de , I updated the branch, can you please rebase? I'm checking this. Thanks!

rebase was done

mmequignon committed:
shopfloor_reception: fix product_uom_qty at move line create

When creating a move line, set product_uom_qty = 0.
Only when such line is posted this value will be aligned with qty_done.

mmequignon committed:
shopfloor_reception: more integration tests

jbaudoux commented on shopfloor_reception/models/shopfloor_menu.py:
```diff
@@ -0,0 +1,28 @@
+class ShopfloorMenu(models.Model):
+    _inherit = "shopfloor.menu"
+
+    allow_select_document_by_product = fields.Boolean(
```
This is the already existing option `scan_location_or_pack_first` that allows to disable scanning product/lot to open a transfer. There is TODO to rename the field as it is misleading

jbaudoux commented on shopfloor_reception/models/shopfloor_menu.py:
```diff
@@ -0,0 +1,28 @@
+    )
+
+    @api.depends("scenario_id")
+    def _compute_pick_pack_same_time_is_possible(self):
```
wrong method name

jbaudoux commented on shopfloor_reception/services/reception.py:
```diff
@@ -839,7 +842,12 @@ def start(self):
+            # only add the handler if allow_select_document_by_product is enabled
+            "product": (
+                self._scan_document__by_product
+                if self.work.menu.allow_select_document_by_product
```
```suggestion
                if not scan_location_or_pack_first
```

jbaudoux commented on shopfloor_reception/services/reception.py:
```diff
@@ -1216,6 +1240,7 @@ def process_without_pack(self, picking_id, selected_line_id, quantity):
         return self._response_for_set_destination(picking, selected_line)
 
     def _post_line(self, selected_line):
+        selected_line.product_uom_qty = selected_line.qty_done
```
I would not do this

mmequignon replied:
when would you do this then?
We agreed that we should set product_uom_qty to 0 at creation, and align it to qty_done at posting.
Am I missing smth?

jbaudoux replied:
I don't think you need to align at posting. If you do it in odoo interface, it will be 0 at creation and remain like that at posting

mmequignon replied:
oh, ok, thanks

jbaudoux commented on shopfloor_reception/services/reception.py:
```diff
@@ -1237,6 +1262,11 @@ def _post_shopfloor_created_line(self, selected_line):
+        # then selected_line is the only one related to this move.
+        # In such case, we must ensure there's another move line with the remaining
+        # quantity to do, so selected_line is extracted in a new move as expected.
+        selected_line._split_partial_quantity()
```
This block requires a `if product_uom_qty:`

jbaudoux commented on shopfloor_reception/services/reception.py:
```diff
@@ -256,13 +256,8 @@ def _scan_line__find_or_create_line(self, picking, move, qty_done=1):
                 )
             )
         )
-        if line:
-            # The line quantity to do needs to correspond to
-            # the remaining quantity to do of its move.
-            line.product_uom_qty = move.product_uom_qty - move.quantity_done
```
The posting is way easier if we always set the product_uom_qty to 0
```
if line:
    line.product_uom_qty = 0
```

mmequignon replied:
We said that if the line wasn't created by shopfloor, we might have product_uom_qty != 0, and we want to keep it like that..
I'm handling this here.
In the cancel action, I unlink move lines with product_uom_qty = 0, and reset the others
https://github.com/OCA/wms/pull/726/files/dd90a411d6c5f84ca0b8c6924297c553da3bd870#diff-57b252eb869558e97f9dbe4859b1f4cafc7db1e67ccdaecc92d2fc38d46a71bdR1167-R1171

jbaudoux commented on shopfloor_reception/services/reception.py:
```diff
@@ -1237,6 +1262,11 @@ def _post_shopfloor_created_line(self, selected_line):
+        # In such case, we must ensure there's another move line with the remaining
+        # quantity to do, so selected_line is extracted in a new move as expected.
+        selected_line._split_partial_quantity()
         new_move = selected_line.move_id.split_other_move_lines(
```
If some other move lines have a product_uom_qty, you need to modify them so that the new_move qty done is not smaller than the remaining move lines product_uom_qty. To prevent this complexity, it is easier if all move lines have a product_uom_qty of 0.
```suggestion
        selected_line.move_id.move_line_ids.filtered("product_uom_qty").product_uom_qty = 0
        new_move = selected_line.move_id.split_other_move_lines(
```

mmequignon replied:
We decided to revert the change where I introduced the "align_qties" method, and to modify the product_uom_qty in the response instead, in order to not alter a move line the user is not working on.
I'm, we should set the qty in `_scan_line__find_or_create_line` every time, whether we create a new line or find one.

jbaudoux replied:
Drop the call to split_other_move_lines and do sth like:
```
if selected_line.qty_done < selected_line.move_id.qty_todo:
            split_move_vals = self._split(selected_line.qty_done)
            split_move = self.create(split_move_vals)
            split_move.move_line_ids = selected_line
            split_move._action_confirm(merge=False)
            split_move._recompute_state()
            split_move._action_assign()
            selected_line.move_id._recompute_state()
```

mmequignon replied:
@jbaudoux can you make another pass please?

mmequignon commented:
@jbaudoux Just checked, and we need the product uom qty to be set on the move line when we split the move.
In split_other_move lines, we use this product_uom_qty to determine the qty to be done on the new move.
Before we post the move though, we can use 0 as a product uom qty, that's ok, I guess.

jbaudoux commented on shopfloor_reception/services/reception.py:
```diff
@@ -768,25 +763,32 @@ def _align_product_uom_qties(self, move):
         rounding = move.product_id.uom_id.rounding
         compare = float_compare(qty_done, qty_todo, precision_rounding=rounding)
-        if compare < 1:  # If qty done <= qty todo, align qty todo on move lines
+        if compare < 1:  # If qty done < qty todo, align qty todo in the response
```
I think we should always compute it. If qty done >= qty todo, you should not return 0 as todo. So I would drop this `if`.

Maybe you could skip `line` in above `qty_done` computation; That would be a `other_lines_qty_done`.
Then below, you don't need to re-add `line.qty_done` to the `remaining_todo`

TDu replied:
If the `if` case is removed, when the users are processing more than the total quantity, how should the computation work ?
Because if the users end up scanning much more that required, we will see negative values in the to do quantity on the front end.

jbaudoux commented on shopfloor_reception/services/reception.py:
```diff
@@ -1237,15 +1262,24 @@ def _post_shopfloor_created_line(self, selected_line):
+        if selected_line.product_uom_qty:
+            selected_line.product_uom_qty = 0
+        move = selected_line.move_id
+        if selected_line.qty_done == move.product_uom_qty:
```
small remark: line and move line could have different uom

jbaudoux commented on shopfloor_reception/services/reception.py:
```diff
@@ -1237,15 +1262,24 @@ def _post_shopfloor_created_line(self, selected_line):
+        move = selected_line.move_id
+        if selected_line.qty_done == move.product_uom_qty:
+            # In case of full quantity, post the initial move
+            selected_line.move_id.extract_and_action_done()
```
I think we are then done here and can return
```suggestion
            return selected_line.move_id.extract_and_action_done()
```

jbaudoux commented on shopfloor_reception/services/reception.py:
```diff
@@ -1237,15 +1262,24 @@ def _post_shopfloor_created_line(self, selected_line):
+        if selected_line.qty_done == move.product_uom_qty:
+            # In case of full quantity, post the initial move
+            selected_line.move_id.extract_and_action_done()
+        split_move_vals = move._split(selected_line.qty_done)
```
small remark: line uom vs move uom

TDu replied:
Fixed

jbaudoux commented on shopfloor_reception/services/reception.py:
```diff
@@ -1237,15 +1262,24 @@ def _post_shopfloor_created_line(self, selected_line):
-        selected_line.move_id.extract_and_action_done()
+        # If user only processed 1/5 and is the only one working on the move,
+        # then selected_line is the only one related to this move.
+        # In such case, we must ensure there's another move line with the remaining
```
```suggestion
        # In such case, we must ensure there's another move with the remaining
```

jbaudoux commented on shopfloor_reception/services/reception.py:
```diff
@@ -1237,15 +1262,24 @@ def _post_shopfloor_created_line(self, selected_line):
+        # then selected_line is the only one related to this move.
+        # In such case, we must ensure there's another move line with the remaining
+        # quantity to do, so selected_line is extracted in a new move as expected.
+        if selected_line.product_uom_qty:
+            selected_line.product_uom_qty = 0
```
Safer to do that on any line having a product_uom_qty. Especially that before your change in this PR, the lines had a computed product_uom_qty.
```suggestion
        lines_with_qty_todo = selected_line.move_id.move_line_ids.filtered("product_uom_qty")
        if lines_with_qty_todo:
          lines_with_qty_todo.product_uom_qty = 0
```

If you have a line with 1/5 and another user creates a line with 1/0 then he cannot post his line unless you change the other initial line

TDu replied:
done

jbaudoux reviewed (changes requested):
Looks promising. I still have a few remarks

TDu commented on shopfloor_reception/services/reception.py:
```diff
@@ -768,25 +763,32 @@ def _align_product_uom_qties(self, move):
+        move_uom = move.product_uom
+        for move_line in move.move_line_ids:
+            # Use move's uom
+            qty_done += move_uom._compute_quantity(
+                move_line.qty_done, move_line.product_uom_id, round=False
+            )
```
Should we not use the move_line_uom
```suggestion
            line_uom = move_line.product_uom_id
            qty_done += line_uom._compute_quantity(
                move_line.qty_done, move_uom.product_uom_id, round=False
            )
```

TDu replied:
Fixed

TDu committed:
shopfloor_reception: Fix product_uom_qty on move line

TDu commented:
@jbaudoux @mt-software-de  Could I get a review update on this one, thanks

mt-software-de commented on shopfloor_reception/data/shopfloor_scenario_data.xml:
```diff
@@ -9,7 +9,8 @@
       "auto_post_line": true,
-      "allow_return": true
+      "allow_return": true,
+      "scan_location_or_pack_first": true
```
Shouldn't scan_location_or_pack_first set by a migration script to True, on a existing reception scenario?

TDu replied:
Not if we want to keep the same behavior for existing implementation.

mt-software-de commented on shopfloor_reception/services/reception.py:
```diff
@@ -256,13 +256,8 @@ def _scan_line__find_or_create_line(self, picking, move, qty_done=1):
-        else:
-            qty_todo_remaining = max(0, move.product_uom_qty - move.quantity_done)
-            values = move._prepare_move_line_vals(quantity=qty_todo_remaining)
+        if not line:
+            values = move._prepare_move_line_vals()
```
Shouldn't there not a kind of lock?
If users work in parrallel, they will get serialize errors.
Because in the method _scan_line__assign_user the picking.user_id will be set and also of course the shopfloor_user_id of the stock.move.line. 
Also i think we should remove the self._assign_user_to_picking from _scan_line__assign_user

If a line was found, shouldn't we change the product_uom_qty to 0. Otherwise _auto_post_line can maybe make some trouble if we work in parallel.

TDu replied:
I think removing `self._assign_user_to_picking `   from _scan_line__assign_user is a good idea because the user assigned to picking is not used in this scenario.
@jbaudoux  ?

About reseting `product_uom_qty` to zero it is probably already done enough in this scenario (Before line is posted, all are reset)

mt-software-de replied:
I mean if a user starts and there is already an unassigned move.line where product_uom_qty is already set. I think it would be safe to set it to 0, then it would be the same as creating a new move.line

mt-software-de replied:
If two users start a the same time with the same product this will cause an serialize access error, because the will both get maybe the same existing move.line if there is one.
And if we are not remove the assign_user_to_picking this will cause also an serialize access error.

TDu replied:
This function has been improved by the commit [shopfloor_reception: improve move line assignation to user](https://github.com/OCA/wms/pull/726/commits/d56d148c0feddef38fe6045852f1e29acb98b2f0)
First to lock a row being assigned to the user.
Second the first implementation missed some sorting and the first line being choosen could have been one without shopfloor_user assigned although a line with the current user assigned does exist.

mt-software-de commented on shopfloor_reception/services/reception.py:
```diff
@@ -1237,15 +1258,34 @@ def _post_shopfloor_created_line(self, selected_line):
-            return
-        # In case of full quantity, post the initial move
-        selected_line.move_id.extract_and_action_done()
+        if lines_with_qty_todo:
+            lines_with_qty_todo.product_uom_qty = 0
```
We should ensure that there is only one line with product_uom_qty on _scan_line__find_or_create_line we should ensure that the product_uom_qty is 0. Otherwise we can find maybe more then one line?

TDu replied:
With this scenario there can be multiple shopfloor users on the same move. Each one will have its own `stock.move.line` assigned. The `product_uom_qty` are kept to zero on all `stock.move.line` until they are being posted. This is to avoid concurrency issues, when users are scanning more than what is required by the `stock.move`.
So it is actually the goal to update all lines with a positive value on `product_uom_qty` back to zero.
All the done line that we would not want to update have been extracted on done.

TDu replied:
I also improved the filter above to exclude on the state as well, just in case and to be more explicit.

mt-software-de replied:
What i mean. In _post_line you are setting the product_uom_qty
This line will be found bei auto_post_line normally it should only find the selected_line,
because other lines from other users should also have set  a qty of 0. 
But in _scan_line__find_or_create_line there could be a existing unassigned move.line returned with a qty set.

Also i think we can remove lines_with_qty_todo and always use the selected line.

mt-software-de replied:
Is it really needed to check for other stock.move.lines with a product_uom_qty > 0?
Because i think every scanned line should have a product_uom_qty of 0, until _post_line is called.

mmequignon replied:
yes, because nobody prevents anybody to modify this from the UI.
Hence, we might have issues if `sum(move.move_line_ids.mapped("product_uom_qty")) > move.product_uom_qty`, when recomputing the state of the move.

mmequignon replied:
We can set product_uom_qty = 0 everywhere in the code, but what really matters, here, is that those are aligned when recomputing the state of the move.
Otherwise the move state moves from assigned to partially available, and is not visible in shopfloor.

mt-software-de replied:
Thx for the explanation. I tested this quick. 
If i'm working with two users. Both are scanning both getting an own move.line. 
The first one creates a new package and sets the the destination and enters the _post_line method.
Within the _auto_post_line the split is happening a new picking is created in state done.
And the move._recompute_state() sets the state to "confirmed". 
So the move state is not anymore 'assigned' as you said. 
What is clear if i am looking at the code. 
https://github.com/odoo/odoo/blob/f572335756ef60402236cf5e4e11c0c843c3879e/addons/stock/models/stock_move.py#L1653
https://github.com/odoo/odoo/blob/f572335756ef60402236cf5e4e11c0c843c3879e/addons/stock/models/stock_move.py#L1659C25-L1659C26
It will only be set to 'assigned' if the move.reserved_availability ( sum of move.lines product_uom_qty) is the same as the move's product_uom_qty, and this isn't the case.

TDu replied:
I could reproduce the issue you mentioned.
My last commit fixes that.

mt-software-de commented on shopfloor_reception/services/reception.py:
```diff
@@ -1237,15 +1258,34 @@ def _post_shopfloor_created_line(self, selected_line):
+        # In such case, we must ensure there's another move with the remaining
+        # quantity to do, so selected_line is extracted in a new move as expected.
+
+        # Always keep the quantity todo at zero, the same is done
+        # in Odoo when move lines are created manually (setting)
+        lines_with_qty_todo = selected_line.move_id.move_line_ids.filtered(
+            "product_uom_qty"
         )
-        if new_move:
-            # A new move is created in case of partial quantity
-            new_move.extract_and_action_done()
-            return
-        # In case of full quantity, post the initial move
-        selected_line.move_id.extract_and_action_done()
+        if lines_with_qty_todo:
+            lines_with_qty_todo.product_uom_qty = 0
+
+        move = selected_line.move_id
+        move_quantity = move.product_uom._compute_quantity(
+            move.product_uom_qty, selected_line.product_uom_id
+        )
+        if selected_line.qty_done == move_quantity:
+            # In case of full quantity, post the initial move
+            return selected_line.move_id.extract_and_action_done()
+        split_move_vals = move._split(selected_line.qty_done)
+        new_move = move.create(split_move_vals)
+        new_move.move_line_ids = selected_line
+        new_move._action_confirm(merge=False)
+        new_move._recompute_state()
+        new_move._action_assign()
+        move._recompute_state()
+        new_move.extract_and_action_done()
```
Also here if we work in parallel we have to add a lock.
Otherwise the recompute_state will run into an error.

mt-software-de replied:
The user should be unassigned from the picking or never set the user on the picking?

TDu replied:
> The user should be unassigned from the picking or never set the user on the picking?

The picking are never selected based on the ` user_id`

TDu replied:
> Also here if we work in parallel we have to add a lock. Otherwise the recompute_state will run into an error.

What kind of error due to concurrent update, example ?

mt-software-de replied:
> > The user should be unassigned from the picking or never set the user on the picking?
> 
> The picking are never selected based on the ` user_id`

I mean if keep the assign of the user on the picking, we should remove it on a split case.

mt-software-de replied:
> > Also here if we work in parallel we have to add a lock. Otherwise the recompute_state will run into an error.
> 
> What kind of error due to concurrent update, example ?

I mean a serialize access error. Because on a split case we always creating a new move which will be set to done. The initial move stays. This means if the users work in parallel and processing the same move which own a move.line one by one. The recompute_state will be always triggered for the same move. And this can run it a serialize access error.

TDu replied:
I added a commit to remove this assignation of the shopfloor user to the picking.

mmequignon replied:
it's done, I resolve the conversation (to make it easier for JE to ignore such discussion)

mt-software-de replied:
But we still need a lock

mt-software-de reviewed (changes requested):
I think we need some locks. Otherwise working in parallel will not work.

TDu commented:
Rebased on 14.0

TDu committed:
sf_reception: remove assigning user to picking

Because many users can work on the same picking at the same time
and the scenario never uses this information.
It is not needed.

TDu committed:
shopfloor_base: improve lock for update action

Add the skip locked option.

TDu committed:
shopfloor_reception: improve move line assignation to user

mt-software-de commented on shopfloor_reception/services/reception.py:
```diff
@@ -1241,15 +1270,35 @@ def _post_shopfloor_created_line(self, selected_line):
+        new_move._action_confirm(merge=False)
+        new_move._recompute_state()
+        new_move._action_assign()
+        move._recompute_state()
```
There should also be a lock.

mt-software-de commented on shopfloor_reception/services/reception.py:
```diff
@@ -1241,15 +1270,35 @@ def _post_shopfloor_created_line(self, selected_line):
+        new_move._action_confirm(merge=False)
+        new_move._recompute_state()
+        new_move._action_assign()
+        move._recompute_state()
```
There should also be a lock. 
```suggestion
        lock = self._actions_for("lock")
        lock.for_update(move)
        move._recompute_state()
```

TDu replied:
Fixed

TDu commented:
@mt-software-de PR updated could you give another review, thank you

jbaudoux commented on shopfloor_reception/services/reception.py:
```diff
@@ -1241,15 +1270,37 @@ def _post_shopfloor_created_line(self, selected_line):
+        new_move._action_confirm(merge=False)
+        new_move._recompute_state()
+        new_move._action_assign()
+        lock = self._actions_for("lock")
+        lock.for_update(move)
```
Should be before the call to _split that modifies the qty to do on the move

jbaudoux replied:
should even be before `if lines_with_qty_todo:`

TDu replied:
I moved it.

jbaudoux commented:
@Tdu Can you squash fixups?

TDu committed:
shopfloor_reception: fix lock on move when done

mt-software-de commented on shopfloor_reception/services/reception.py:
```diff
@@ -248,27 +245,37 @@ def _select_document_from_lot(self, lot):
+                if lock.for_update(move_line, skip_locked=True):
+                    line = move_line
+                    break
+        if not line:
+            values = move._prepare_move_line_vals()
             line = self.env["stock.move.line"].create(values)
```
I think we should ensure that the product_uom_qty is always 0. 

```suggestion
        if not line:
            values = move._prepare_move_line_vals()
            line = self.env["stock.move.line"].create(values)
        else:
            line.product_uom_qty = 0
```

jbaudoux replied:
We do this on the post for backward compatibility with previous implementation (lines that would already be there with a qty != 0). Also if you have only one line pre-generated by odoo, you have a qty != 0 and this is not causing any issue.
When posting, it is required we care other lines (if any) are well set to 0 otherwise you cannot post.

mmequignon replied:
https://github.com/OCA/wms/pull/726#discussion_r1394116955

jbaudoux commented:
@mt-software-de I think everything as been addressed now. I have not re-tested. Can you update your review?

mt-software-de commented:
> @mt-software-de I think everything as been addressed now. I have not re-tested. Can you update your review?
Not everything. Could you have a look at my last review?

mmequignon commented:
> > @mt-software-de I think everything as been addressed now. I have not re-tested. Can you update your review?
> > Not everything. Could you have a look at my last review?

Thanks for your insights! :pray: 

However, this PR was meant to revert a specific change, and I believe we went too far.
Now that this looks (IMO) good, we should focus on merging this, and open a new PR later if needed.
This will help to have kind of atomic PRs, and also make the merging faster.
3 months for what was initially supposed to be a "simple" revert change is way too long.
It doesn't help maintaining those pending merges.

mt-software-de commented:
> > > @mt-software-de I think everything as been addressed now. I have not re-tested. Can you update your review?
> > > Not everything. Could you have a look at my last review?
> 
> Thanks for your insights! 🙏
> 
> However, this PR was meant to revert a specific change, and I believe we went too far. Now that this looks (IMO) good, we should focus on merging this, and open a new PR later if needed. This will help to have kind of atomic PRs, and also make the merging faster. 3 months for what was initially supposed to be a "simple" revert change is way too long. It doesn't help maintaining those pending merges.

I understand you, but i have unanswered questions to code which was introduced within this PR, maybe i am missing something. But i think this should be answered?

TDu committed:
shopfloor_reception: fix auto post option

When 2 users work on the same move and one user completes his picking,
The move containing the quantity left to do could not be switch back
to assigned.

TDu commented:
Rebased

TDu commented:
Rebased

mt-software-de commented on shopfloor_reception_mobile/static/src/scenario/reception.js:
```diff
@@ -153,6 +153,7 @@ const Reception = {
                         <v-row align="center">
                             <v-col class="text-center" cols="12">
                                 <btn-back/>
+                                <cancel-button/>
```
```suggestion
                        </v-col>
                        <v-col class="text-center" cols="12">
                                <cancel-button/>
```

mt-software-de replied:
A own v-col would be great i think. 
![image](https://github.com/OCA/wms/assets/78783475/e628ec3f-c05a-4fc1-87ee-fb4ef2ed8722)

With suggestion, it would look like this. 
![image](https://github.com/OCA/wms/assets/78783475/3dd8ab47-29fa-4a05-9a25-f23a051ace95)

jbaudoux replied:
Do we need both back & cancel ?

TDu replied:
I don't know, but
the Cancel button will call the back end to cancel the done quantity on the current line (line will be removed).
the Back button goes back to the previous screen without calling the backend.

TDu replied:
I have done the change to have the button spaced out.

jbaudoux replied:
In my opinion, it should be either cancel or back. So cancel should replace the back button. Maybe @mmequignon can clarify the intention here?

mt-software-de reviewed (approved):
LGTM now :-). Thanks

jbaudoux commented:
/ocabot merge minor
